File: backend/handlers/issue_trend_handler.py
from datetime import datetime
from sqlalchemy import func
from sqlalchemy.orm import Session

from handlers.base import WidgetDataHandler
from models import Issue, Member, Department, Project
from schemas import WidgetQueryRequest, WidgetDataResponse, SeriesData, SeriesPoint, ResultMeta

MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun",
          "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]


class IssueTrendHandler(WidgetDataHandler):
    @property
    def source_key(self) -> str:
        return "issue_created_resolved_trend"

    def query(self, request: WidgetQueryRequest, db: Session) -> WidgetDataResponse:
        f = request.filters
        year          = (f.year if f else None) or datetime.now().year
        department_id = f.departmentId if f else None

        project = db.query(Project).filter(Project.key == request.projectKey).first()
        if not project:
            raise ValueError(f"Project '{request.projectKey}' not found")

        def monthly_counts(date_col, extra_filter=None):
            q = (
                db.query(
                    func.strftime("%m", date_col).label("month"),
                    func.count().label("cnt"),
                )
                .join(Issue.assignee)
                .join(Member.department)
                .filter(
                    Issue.project_id == project.id,
                    func.strftime("%Y", date_col) == str(year),
                )
            )
            if extra_filter is not None:
                q = q.filter(extra_filter)
            if department_id:
                q = q.filter(Department.id == department_id)
            return {int(r.month): r.cnt for r in q.group_by(func.strftime("%m", date_col)).all()}

        created_rows  = monthly_counts(Issue.created_at)
        resolved_rows = monthly_counts(Issue.resolved_at, Issue.resolved_at.isnot(None))

        # Build cumulative series
        c_cum = r_cum = 0
        c_data, r_data = [], []
        for m in range(1, 13):
            c_cum += created_rows.get(m, 0)
            r_cum += resolved_rows.get(m, 0)
            c_data.append(SeriesPoint(x=MONTHS[m - 1], y=c_cum))
            r_data.append(SeriesPoint(x=MONTHS[m - 1], y=r_cum))

        return WidgetDataResponse(
            dataSource=self.source_key,
            type="line_chart",
            series=[
                SeriesData(name="Created",  data=c_data),
                SeriesData(name="Resolved", data=r_data),
            ],
            meta=ResultMeta(computedAt=datetime.utcnow().isoformat(), fromCache=False, totalRows=12),
        )

File: backend-java/src/main/java/com/dashboard/handler/IssueTrendHandler.java
package com.dashboard.handler;

import com.dashboard.dto.IssueTrendResponse;
import com.dashboard.dto.IssueTrendResponse.SeriesData;
import com.dashboard.dto.IssueTrendResponse.SeriesPoint;
import jakarta.persistence.EntityManager;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.*;

@Component
public class IssueTrendHandler {

    private static final String[] MONTHS = {
        "Jan","Feb","Mar","Apr","May","Jun",
        "Jul","Aug","Sep","Oct","Nov","Dec"
    };

    public IssueTrendResponse query(String projectKey, Integer year, Long departmentId,
                                    EntityManager em) {
        if (year == null) year = LocalDateTime.now().getYear();
        long deptParam = departmentId != null ? departmentId : -1L;

        String createdSql = """
            SELECT CAST(EXTRACT(MONTH FROM i.created_at) AS INTEGER) AS month, COUNT(*) AS cnt
            FROM issues i
            LEFT JOIN members m ON i.assignee_id = m.id
            LEFT JOIN departments d ON m.department_id = d.id
            WHERE i.project_id = (SELECT id FROM projects WHERE key = :key)
              AND EXTRACT(YEAR FROM i.created_at) = :year
              AND (:deptId = -1 OR d.id = :deptId)
            GROUP BY month
            ORDER BY month
            """;

        String resolvedSql = """
            SELECT CAST(EXTRACT(MONTH FROM i.resolved_at) AS INTEGER) AS month, COUNT(*) AS cnt
            FROM issues i
            LEFT JOIN members m ON i.assignee_id = m.id
            LEFT JOIN departments d ON m.department_id = d.id
            WHERE i.project_id = (SELECT id FROM projects WHERE key = :key)
              AND i.resolved_at IS NOT NULL
              AND EXTRACT(YEAR FROM i.resolved_at) = :year
              AND (:deptId = -1 OR d.id = :deptId)
            GROUP BY month
            ORDER BY month
            """;

        Map<Integer, Integer> createdMap  = runMonthQuery(em, createdSql,  projectKey, year, deptParam);
        Map<Integer, Integer> resolvedMap = runMonthQuery(em, resolvedSql, projectKey, year, deptParam);

        List<SeriesPoint> cData = new ArrayList<>();
        List<SeriesPoint> rData = new ArrayList<>();
        int cCum = 0, rCum = 0;
        for (int m = 1; m <= 12; m++) {
            cCum += createdMap.getOrDefault(m, 0);
            rCum += resolvedMap.getOrDefault(m, 0);
            cData.add(SeriesPoint.builder().x(MONTHS[m - 1]).y(cCum).build());
            rData.add(SeriesPoint.builder().x(MONTHS[m - 1]).y(rCum).build());
        }

        return IssueTrendResponse.builder()
                .series(List.of(
                    SeriesData.builder().name("Created").data(cData).build(),
                    SeriesData.builder().name("Resolved").data(rData).build()
                ))
                .build();
    }

    @SuppressWarnings("unchecked")
    private Map<Integer, Integer> runMonthQuery(EntityManager em, String sql,
                                                String projectKey, int year, long deptId) {
        var query = em.createNativeQuery(sql)
                      .setParameter("key", projectKey)
                      .setParameter("year", year)
                      .setParameter("deptId", deptId);

        Map<Integer, Integer> result = new HashMap<>();
        for (Object[] row : (List<Object[]>) query.getResultList()) {
            result.put(((Number) row[0]).intValue(), ((Number) row[1]).intValue());
        }
        return result;
    }
}
